astra/src/main/java/com/slack/astra/clusterManager/ClusterMonitorService.java

```diff
@@ -54,9 +166,147 @@ public ClusterMonitorService(
         "cached_service_nodes_size", datasetMetadataStore, store -> store.listSync().size());
   }
 
+  private void updatePerPodMetrics() {
+    Thread.ofVirtual()
+        .start(
+            () ->
+                liveChunksPerPod.register(
+                    cacheNodeMetadataStore.listSync().stream()
+                        .map(
+                            cacheNodeMetadata ->
+                                MultiGauge.Row.of(
+                                    Tags.of(Tag.of("pod", cacheNodeMetadata.getName())),
+                                    cacheNodeIdToLiveChunksPerPod.computeIfAbsent(
+                                        cacheNodeMetadata.getName(),
+                                        (_) ->
+                                            new AtomicInteger(
+                                                calculateLiveChunks(cacheNodeMetadata.id)))))
+                        .collect(Collectors.toUnmodifiableList()),
+                    true));
+
+    Thread.ofVirtual()
+        .start(
+            () ->
+                freeSpacePerPod.register(
+                    cacheNodeMetadataStore.listSync().stream()
+                        .map(
+                            cacheNodeMetadata ->
+                                MultiGauge.Row.of(
+                                    Tags.of(Tag.of("pod", cacheNodeMetadata.getName())),
+                                    cacheNodeIdToFreeSpaceBytes.computeIfAbsent(
+                                        cacheNodeMetadata.getName(),
+                                        (_) ->
+                                            new AtomicLong(
+                                                calculateFreeSpaceForPod(cacheNodeMetadata.id)))))
+                        .collect(Collectors.toUnmodifiableList()),
+                    true));
+  }
+
+  private void cacheNodeAssignmentListener(CacheNodeAssignment assignment) {
+    // only update metrics if something has become live
+    if (assignment.state == Metadata.CacheNodeAssignment.CacheNodeAssignmentState.LIVE) {
+      LOG.info("updating per pod statistics due to assignment change");
+      Thread.ofVirtual().start(() -> updateLiveChunksPerPod(assignment));
```

Comment:
Same here, is there a reason to use multiple virtual threads?

Reply:
When we redeploy cache nodes, there ends up being a huge barrage of ZK events as assignments for the old cache nodes are deleted. This is partially what was causing the manager to crash previously. The virtual thread prevents the manager from getting overwhelmed and killed from failing health checks. 

I'm honestly not a huge fan of this approach; I was thinking it might be worth it at this point to make this service a scheduled one and add some debouncing for metric updating? Open to any other ideas!